src/env/mod.rs

```diff
@@ -10,11 +11,57 @@ use rng256::Rng256;
 pub mod test;
 pub mod tock;
 
+pub enum SendOrRecvStatus {
+    Timeout,
+    Sent,
+    Received,
+}
+
+pub struct SendOrRecvError;
+
+pub type SendOrRecvResult = Result<SendOrRecvStatus, SendOrRecvError>;
+
+pub trait CtapHidChannel {
+    fn send_or_recv_with_timeout(
+        &mut self,
+        buf: &mut [u8; 64],
+        timeout: CtapDuration,
+    ) -> SendOrRecvResult;
+}
+
+pub enum UserPresenceStatus {
+    Confirmed,
+    Declined,
+    Timeout,
+}
+
+pub type UserPresenceResult = Result<UserPresenceStatus, Ctap2StatusCode>;
+
 pub trait UserPresence {
-    /// Blocks for user presence.
+    /// Called at the beginning of user presence checking process.
+    fn user_presence_check_init(&mut self, _channel: Channel) {}
+
+    /// Implements a wait for user presence confirmation or rejection.
+    fn wait_for_user_presence_with_timeout(
+        &mut self,
+        _channel: Channel,
+        _timeout: CtapDuration,
+    ) -> UserPresenceResult {
+        Ok(UserPresenceStatus::Confirmed)
+    }
+
+    /// Called at the end of user presence checking process.
+    fn user_presence_check_complete(&mut self, _result: &UserPresenceResult) {}
+
+    /// Short-circuit implementation for fast user presence checking.
     ///
-    /// Returns an error in case of timeout or keepalive error.
-    fn check(&mut self, channel: Channel) -> Result<(), Ctap2StatusCode>;
+    /// Default algorithm for user presence checking waits for user action, while sending keepalive
+    /// packets to the user agent. Implementations may use this function to override this behavior
+    /// and return Some(Result<...>) to pass it immediately to the caller wanting to check user
+    /// presence.
+    fn quick_check(&mut self, _channel: Channel) -> Option<Result<(), Ctap2StatusCode>> {
```

Comment:
I'm not convinced we need this long-term. I see 2 options:
- It's possible to implement the test behavior with the init/wait/complete API. In which case we should remove this.
- It's not (yet) possible to implement the test behavior with the existing API. In which case we can keep this and explicitly mention that it's for tests and temporary.